### PR TITLE
Tb le added editor from Genie

### PR DESCRIPTION
**Pull request by TomasBlazek1**
All new files are from Genie. There were some minor modifications, but I didn't find a simple way how to keep them visible.

### CODE REVIEW

File: src/LayerEditor/ui/tools/selection.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        self._selected = []
+
+    def select_item(self, item):
+        self._selected.clear()
+        self.select_add_item(item)
+        self._diagram.update()
+
+        self._diagram.selection_changed.emit()
+
+    def select_add_item(self, item):
```
Comment (jbrezmorf) [inline src/LayerEditor/ui/tools/selection.py:13]: ? rename select_toggle_item

File: src/LayerEditor/ui/diagram_editor/diagram.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+class Diagram(QtWidgets.QGraphicsScene):
+    selection_changed = QtCore.pyqtSignal()
+    # selection has changed
+
+    regionsUpdateRequired = QtCore.pyqtSignal()
+
+    def __init__(self, parent):
+        rect = QtCore.QRectF(-622500, 1128600, 400, 500)
+        super().__init__(rect, parent)
+        self.points = {}
```
Comment (jbrezmorf) [inline src/LayerEditor/ui/diagram_editor/diagram.py:26]: Purpose of these maps?

File: src/LayerEditor/ui/diagram_editor/diagram.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+        self.polygons = {}
+
+        self.regions = Regions(self)
+
+        self.last_point = None
+        self.aux_pt, self.aux_seg = self.create_aux_segment()
+        self.hide_aux_line()
+
+        self._zoom_value = 1.0
+        self.selection = Selection(self)
```
Comment (jbrezmorf) [inline src/LayerEditor/ui/diagram_editor/diagram.py:37]: Is selection specific to single diagram or is it independent of the diagram? I think it is the second case, so the selection should probably not be part of the diagram.

File: src/LayerEditor/ui/diagram_editor/diagram.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+        if type(shape) is GsPoint:
+            return 1, shape.pt.id
+
+        elif type(shape) is GsSegment:
+            return 2, shape.segment.id
+
+        elif type(shape) is GsPolygon:
+            return 3, shape.polygon_data.id
+
+    def undo(self):
```
Comment (jbrezmorf) [inline src/LayerEditor/ui/diagram_editor/diagram.py:320]: Undo and redo operations should be part of main window.

File: src/LayerEditor/ui/data/regions.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    def delete_region(self, id):
+        del self.regions[id]
+
+    def get_region_names(self):
+        return [reg.name for reg in self.regions.values()]
+
+    def get_shape_region(self, shape_key):
+        dim, shape_id = shape_key
+        attr = None
+        if dim == 1:
```
Comment (jbrezmorf) [inline src/LayerEditor/ui/data/regions.py:23]: Can use `decomposition.shapes[dim][shape_id].attr`.

File: src/LayerEditor/ui/data/regions.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
+from LayerEditor.ui.data.region import Region
+
+
+class Regions:
+    def __init__(self, diagram):
+        self.regions = {Region.none.id: Region.none}
+        self._diagram = diagram
```
Comment (jbrezmorf) [inline src/LayerEditor/ui/data/regions.py:7]: Is the reference to the whole diagram necessary? Possibly keep just reference to the decomposition.

File: src/LayerEditor/ui/data/regions.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        elif dim == 3:
+            attr = self._diagram.decomposition.polygons[shape_id].attr
+
+        if attr is None:
+            attr = Region.none.id
+
+        return attr
+
+    def get_common_region(self):
+        selected = self._diagram.selection._selected
```
Comment (jbrezmorf) [inline src/LayerEditor/ui/data/regions.py:36]: Rather pass the selection as a parameter.

File: src/LayerEditor/ui/data/regions.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        return r_id
+
+    def set_region(self, dim, shape_id, reg_id):
+        if dim != self.regions[reg_id].dim:
+            return False
+
+        if reg_id is None:
+            reg_id = Region.none.id
+
+        if dim == 1:
```
Comment (jbrezmorf) [inline src/LayerEditor/ui/data/regions.py:52]: use `decomposition.shapes`

File: src/LayerEditor/ui/data/regions.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            self._diagram.decomposition.segments[shape_id].attr = reg_id
+        elif dim == 3:
+            self._diagram.decomposition.polygons[shape_id].attr = reg_id
+
+        return True
+
+    def is_region_used(self, reg_id):
+        dim = self.regions[reg_id].dim
+        elements = []
+        if dim == 1:
```
Comment (jbrezmorf) [inline src/LayerEditor/ui/data/regions.py:64]: See previous note.

**Review by jbrezmorf — COMMENTED** [review]
- comment purpose of data attributes
- comment purpose of classes
- can ommit purpose of individual methods for now
- in future we should separate modeal and view classes by directories, now in order to minimise changes we should mark model classes in comments `MODEL`. These should have all modification operations unduable.

File: src/LayerEditor/ui/diagram_editor/diagram.py (hunk 2 above)
Comment (TomasBlazek1) [reply src/LayerEditor/ui/diagram_editor/diagram.py:26]: Maybe they could be lists, but I found at least one instance when the graphics object is referenced by id.

File: src/LayerEditor/ui/diagram_editor/diagram.py (hunk 3 above)
Comment (TomasBlazek1) [reply src/LayerEditor/ui/diagram_editor/diagram.py:37]: Solved in branch TB_LE_inicialization_and_IO

File: src/LayerEditor/ui/data/regions.py (hunk 5 above)
Comment (TomasBlazek1) [reply src/LayerEditor/ui/data/regions.py:23]: Used in branch TB_LE_inicialization_and_IO

File: src/LayerEditor/ui/data/regions.py (hunk 6 above)
Comment (TomasBlazek1) [reply src/LayerEditor/ui/data/regions.py:7]: Deleted in branch TB_LE_inicialization_and_IO

**Pull request merged**